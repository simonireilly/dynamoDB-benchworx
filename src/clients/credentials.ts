import { defaultProvider } from '@aws-sdk/credential-provider-node'

const credentials = new defaultProvider()
