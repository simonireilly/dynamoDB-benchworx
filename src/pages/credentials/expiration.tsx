import React, { ReactElement, useContext, useEffect, useState } from "react";
import { ElectronStore } from "@src/contexts/electron-context";
import { Chip } from "@material-ui/core";

const calculateSecondsLeft = (date: Date): number => {
  if (date) return Math.abs(date.getTime() - new Date().getTime()) / 1000;
};

export const Expiration = (): ReactElement => {
  const { credentials } = useContext(ElectronStore);

  const [secondsLeft, setSecondsLeft] = useState<number>(
    calculateSecondsLeft(credentials.expiration)
  );

  useEffect(() => {
    const timer = setTimeout(() => {
      setSecondsLeft(calculateSecondsLeft(credentials.expiration));
    }, 500);

    return clearTimeout(timer);
  });

  return credentials.expiration ? (
    <Chip
      label={`Session expiry: ${secondsLeft}`}
      color="secondary"
      variant="outlined"
    />
  ) : (
    <></>
  );
};
