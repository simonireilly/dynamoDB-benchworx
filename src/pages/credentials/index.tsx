import {
  FormControl,
  FormControlLabel,
  FormLabel,
  Radio,
  RadioGroup,
} from "@material-ui/core";
import React, { ReactElement, useContext, useEffect, useState } from "react";
import { ElectronStore } from "../contexts/electron-context";

export const Credentials = (): ReactElement => {
  const { aws } = useContext(ElectronStore);

  const [credentials, setCredentials] = useState({});

  useEffect(() => {
    const fetchCredentials = async () => {
      try {
        const configuration = await aws.listCredentials();
        setCredentials(configuration);
      } catch (e) {
        console.error(e);
      }
    };

    fetchCredentials();
  }, []);

  const handleChange = () => {
    return "";
  };

  return (
    <>
      <FormControl component="fieldset">
        <FormLabel component="legend">Gender</FormLabel>
        <RadioGroup
          aria-label="gender"
          name="gender1"
          value={credentials}
          onChange={handleChange}
        >
          <FormControlLabel value="female" control={<Radio />} label="Female" />
          <FormControlLabel value="male" control={<Radio />} label="Male" />
          <FormControlLabel value="other" control={<Radio />} label="Other" />
          <FormControlLabel
            value="disabled"
            disabled
            control={<Radio />}
            label="(Disabled option)"
          />
        </RadioGroup>
      </FormControl>
      <pre>
        <h1>AWS Credentials available for configuration</h1>;
        <pre>{JSON.stringify(credentials, null, 2)}</pre>
      </pre>
    </>
  );
};
