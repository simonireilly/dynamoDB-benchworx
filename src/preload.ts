console.info("Preloading node modules");

import { contextBridge } from "electron";

import {
  loadSharedConfigFiles,
  SharedConfigFiles,
} from "@aws-sdk/shared-ini-file-loader";

declare global {
  interface Window {
    aws: {
      listCredentials: () => Promise<SharedConfigFiles>;
    };
  }
}

contextBridge.exposeInMainWorld("aws", {
  listCredentials: () => loadSharedConfigFiles(),
});
