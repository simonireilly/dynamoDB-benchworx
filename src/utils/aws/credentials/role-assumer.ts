import { STSClient, AssumeRoleCommand } from "@aws-sdk/client-sts";
import { Credentials } from "@aws-sdk/types";
import { AssumeRoleParams } from "@aws-sdk/credential-provider-ini";

// Assume a role using the source credentials provided
export const roleAssumer = async (
  sourceCredentials: Credentials,
  params: AssumeRoleParams
): Promise<Credentials> => {
  const client = new STSClient({
    region: "us-east-1",
    credentials: sourceCredentials,
  });

  const command = new AssumeRoleCommand(params);
  const response = await client.send(command);

  return {
    expiration: response.Credentials.Expiration,
    accessKeyId: response.Credentials.AccessKeyId,
    secretAccessKey: response.Credentials.SecretAccessKey,
    sessionToken: response.Credentials.SessionToken,
  };
};
