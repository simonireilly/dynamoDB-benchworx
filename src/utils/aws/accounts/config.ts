// Get Credentials from the local INI files and organize into types for sign in
// which are safe for persistance and use inside the DOM

import { PreloaderResponse } from "@src/preload";
import {
  loadSharedConfigFiles,
  Profile,
  SharedConfigFiles,
} from "@aws-sdk/shared-ini-file-loader";

export type SafeProfile = {
  profile: string;
  mfa: boolean;
  assumeRole: boolean;
  region: string;
  endpoint?: string;
};

export const listAwsConfig = async (): Promise<
  PreloaderResponse<SafeProfile[]>
> => {
  const config = await loadSharedConfigFiles();

  if (config) {
    const data = Object.entries(mergeCredentialsAndConfig(config)).map(
      safeConfigConstructor
    );

    console.info(data);

    return {
      type: "success",
      data,
      message: "Fetched local roles from INI files in ~/.aws",
      details: `Configurations found: ${data.length}`,
    };
  } else {
    return {
      type: "error",
      data: null,
      message: "Failed to fetch local roles from INI files in ~/.aws/",
      details: "",
    };
  }
};

const mergeCredentialsAndConfig = (config: SharedConfigFiles) => {
  const profiles: { [key: string]: Profile } = {};

  const credentialProfiles = Object.entries(config.credentialsFile).reduce(
    (acc, [key, values]) => {
      acc[key] = {
        ...values,
        ...config.configFile[key],
      };

      return acc;
    },
    profiles
  );

  const completeProfiles = Object.entries(config.configFile).reduce(
    (acc, [key, values]) => {
      acc[key] = {
        ...acc[key],
        ...values,
      };

      return acc;
    },
    credentialProfiles
  );

  return completeProfiles;
};

const safeConfigConstructor = (entry: [string, Profile]): SafeProfile => {
  const [profile, data = {}] = entry;
  const { aws_access_key_id, aws_secret_access_key, ...safeData } = data;

  const currentProfile: SafeProfile = {
    profile,
    mfa: Boolean(safeData.mfa_serial),
    assumeRole: Boolean(safeData.role_arn),
    region: safeData.region,
  };

  if (safeData.endpoint) currentProfile.endpoint = safeData.endpoint;

  return currentProfile;
};
