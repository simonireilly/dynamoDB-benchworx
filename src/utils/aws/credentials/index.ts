import { fromIni } from "@aws-sdk/credential-provider-ini";

import { roleAssumer } from "@src/utils/aws/credentials/role-assumer";
import { Credentials } from "@aws-sdk/types";
import { PreloaderResponse } from "@src/preload";

type Props = {
  profile: string;
  mfaCode: string;
};

const inMemoryCache: {
  [key: string]: Credentials;
} = {};

export const authenticator = async ({
  profile,
  mfaCode,
}: Props): Promise<PreloaderResponse<void>> => {
  const credentials = fromIni({
    profile,
    roleAssumer,
    mfaCodeProvider: (mfaSerial: string): Promise<string> => {
      return new Promise((resolve) => resolve(mfaCode));
    },
  });

  try {
    inMemoryCache[profile] = await credentials();

    return {
      type: "success",
      message: `Authenticated for ${profile}`,
      details: null,
      data: null,
    };
  } catch (e) {
    return {
      type: "error",
      message: `Failed to authenticate for ${profile}`,
      details: e.message,
      data: null,
    };
  }
};

export const fetchCredentials = (profile: string): Credentials => {
  return inMemoryCache[profile];
};
