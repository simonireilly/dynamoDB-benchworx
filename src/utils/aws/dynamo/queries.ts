import { PreloaderResponse } from "@src/preload";

import {
  DescribeTableCommand,
  DescribeTableCommandOutput,
  DynamoDBClient,
  ListTablesCommand,
  ListTablesCommandInput,
  ListTablesCommandOutput,
} from "@aws-sdk/client-dynamodb";

import {
  DynamoDBDocument,
  ScanCommandInput,
  ScanCommandOutput,
} from "@aws-sdk/lib-dynamodb";
import { fetchCredentials } from "@src/utils/aws/credentials";

const clientConstructor = async (profile: string, region: string) => {
  const credentials = await fetchCredentials(profile);

  const client = new DynamoDBClient({
    region,
    credentials,
    logger: console,
  });

  return client;
};

const fetchAllTables = async (
  client: DynamoDBClient,
  previousResult?: ListTablesCommandOutput
): Promise<ListTablesCommandOutput> => {
  let TableNames: string[] = [];
  const params: ListTablesCommandInput = {
    Limit: 100,
  };

  if (previousResult?.LastEvaluatedTableName)
    params.ExclusiveStartTableName = previousResult.LastEvaluatedTableName;

  const command = new ListTablesCommand(params);
  const results = await client.send(command);

  if (results.TableNames.length === 100) {
    TableNames = [...TableNames, ...results.TableNames];
    const result = await fetchAllTables(client, results);
    TableNames = [...TableNames, ...result.TableNames];
    return {
      ...result,
      TableNames,
    };
  } else {
    return results;
  }
};

// Assume profile and list tables
export const listTables = async (
  profile: string,
  region: string
): Promise<PreloaderResponse<ListTablesCommandOutput>> => {
  let result;

  try {
    const client = await clientConstructor(profile, region);

    result = await fetchAllTables(client);

    return {
      type: "success",
      data: result,
      message: `Fetched list of tables from dynamo with ${profile}`,
      details: `Table count: ${result.TableNames.length}`,
    };
  } catch (e) {
    return {
      type: "error",
      data: null,
      message: `Unable to list tables for profile: ${profile}`,
      details: e.message,
    };
  }
};

export const describeTable = async (
  profile: string,
  region: string,
  tableName: string
): Promise<PreloaderResponse<DescribeTableCommandOutput>> => {
  let result;

  try {
    const client = await clientConstructor(profile, region);

    result = await client.send(
      new DescribeTableCommand({ TableName: tableName })
    );

    return {
      type: "success",
      data: result,
      message: `Fetched table schema for ${tableName}`,
      details: `Table item count: ${result.Table.ItemCount}`,
    };
  } catch (e) {
    return {
      type: "error",
      data: null,
      message: `Unable to list tables for profile: ${profile}`,
      details: e.message,
    };
  }
};

export const scan = async (
  profile: string,
  region: string,
  options: ScanCommandInput
): Promise<PreloaderResponse<ScanCommandOutput>> => {
  let result;

  try {
    const client = await clientConstructor(profile, region);

    const documentClient = DynamoDBDocument.from(client);

    result = await documentClient.scan(options);

    return {
      type: "success",
      data: result,
      message: `Scanned table: ${options.TableName}`,
      details: `Scan count: ${result.Count}`,
    };
  } catch (e) {
    return {
      type: "error",
      data: null,
      message: `Scan operation failed: ${options.TableName}`,
      details: e.message,
    };
  }
};
