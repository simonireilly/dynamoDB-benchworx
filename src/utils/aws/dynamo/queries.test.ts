import { mocked } from "ts-jest/utils";
import { describeTable, listTables, scan } from "./queries";
import nock from "nock";
import {
  ListTablesCommandInput,
  ListTablesCommandOutput,
} from "@aws-sdk/client-dynamodb";
import { fetchCredentials } from "@src/utils/aws/credentials";
import { listAwsConfig } from "@src/utils/aws/accounts/config";
import { describeTableResponse } from "@fixtures/index";
import { PreloaderResponse } from "@src/preload";

jest.spyOn(global.console, "info");

jest.mock("@src/utils/aws/credentials");
const mockedFetchCredentials = mocked(fetchCredentials);

jest.mock("@src/utils/aws/accounts/config");
const mockedListAwsConfig = mocked(listAwsConfig);

describe("Queries", () => {
  beforeEach(() => {
    jest.resetAllMocks();
    nock.disableNetConnect();

    mockedFetchCredentials.mockResolvedValueOnce({
      accessKeyId: "test",
      secretAccessKey: "secret",
    });

    mockedListAwsConfig.mockResolvedValueOnce({
      type: "success",
      data: [
        { profile: "cgu", mfa: false, assumeRole: false, region: "eu-west-1" },
      ],
      message: null,
      details: null,
    });
  });

  describe("listTables", () => {
    it("lists all available tables", async () => {
      const TableNames = ["Table-1", "Table-2", "Table-3"];
      // Mock post Request
      nock("https://dynamodb.eu-west-1.amazonaws.com").post("/").reply(200, {
        TableNames,
      });

      // Act
      const result = await listTables("cgu", "eu-west-1");

      // assert
      expect(result.data.TableNames).toEqual(
        expect.arrayContaining(TableNames)
      );
    });

    it("recursively fetches tables when greater than 100 tables found", async () => {
      const TableNames1 = [...Array.from({ length: 100 }, (x, i) => ++i)].map(
        (key: number) => `Table-${key}`
      );
      const TableNames2 = [
        ...Array.from({ length: 100 }, (x, i) => ++i + 100),
      ].map((key: number) => `Table-${key}`);
      const TableNames3 = ["Table-201"];

      nock("https://dynamodb.eu-west-1.amazonaws.com")
        .post("/")
        .reply(200, {
          TableNames: TableNames1,
          LastEvaluatedTableName: "Table-100",
          $metadata: {},
        } as ListTablesCommandOutput);

      nock("https://dynamodb.eu-west-1.amazonaws.com")
        .post("/", {
          Limit: 100,
          ExclusiveStartTableName: "Table-100",
        })
        .reply(200, {
          TableNames: TableNames2,
          LastEvaluatedTableName: "Table-200",
          $metadata: {},
        } as ListTablesCommandOutput);

      nock("https://dynamodb.eu-west-1.amazonaws.com")
        .post("/", {
          Limit: 100,
          ExclusiveStartTableName: "Table-200",
        } as ListTablesCommandInput)
        .reply(200, {
          TableNames: TableNames3,
          LastEvaluatedTableName: "Table-201",
          $metadata: {},
        } as ListTablesCommandOutput);

      // Act
      const result = await listTables("cgu", "eu-west-1");

      // assert
      expect(result.data.TableNames.length).toEqual(201);
      expect(result.data.TableNames).toEqual(
        expect.arrayContaining([...TableNames1, ...TableNames2, ...TableNames3])
      );
    });
  });

  describe("scan", () => {
    it("calls scan command on the document client", async () => {
      const Items = [
        {
          ReplyDateTime: {
            S: "2019-10-31 11:27:17",
          },
          Message: {
            S: "DynamoDB Thread 1 Reply 2 text",
          },
          PostedBy: {
            S: "User B",
          },
          Id: {
            S: "Amazon DynamoDB#DynamoDB Thread 1",
          },
        },
      ];
      // Mock post Request
      nock("https://dynamodb.eu-west-1.amazonaws.com").post("/").reply(200, {
        Count: 1,
        Items,
        ScannedCount: 1,
      });

      // Act
      const result = await scan("cgu", "eu-west-1", {
        TableName: "test-table",
      });

      // assert
      expect(result.data.Items).toEqual(
        expect.arrayContaining([
          expect.objectContaining({
            ReplyDateTime: "2019-10-31 11:27:17",
            Message: "DynamoDB Thread 1 Reply 2 text",
            PostedBy: "User B",
            Id: "Amazon DynamoDB#DynamoDB Thread 1",
          }),
        ])
      );
    });
  });

  describe("describeTable", () => {
    it("returns the information for the table", async () => {
      // Mock post Request
      nock("https://dynamodb.eu-west-1.amazonaws.com")
        .post("/")
        .reply(200, describeTableResponse);

      // Act
      const result = await describeTable("cgu", "eu-west-1", "test-table");

      // assert
      expect(result.data).toMatchSnapshot();
    });
  });

  describe("error handling", () => {
    let result: PreloaderResponse<any>;
    beforeEach(() => {
      // Mock post Request
      nock("https://dynamodb.eu-west-1.amazonaws.com").post("/").reply(404);
    });

    afterEach(() => {
      // assert
      expect(result.type).toEqual("error");
    });
    it("listTables", async () => {
      // Act
      result = await listTables("cgu", "eu-west-1");
    });
    it("scan", async () => {
      // Act
      result = await scan("cgu", "eu-west-1", {
        TableName: "test-table",
      });
    });
    it("query", async () => {
      // Act
      result = await query("cgu", "eu-west-1", {
        TableName: "test-table",
      });
    });
    it("describeTable", async () => {
      // Act
      result = await describeTable("cgu", "eu-west-1", "test-table");
    });
  });
});
