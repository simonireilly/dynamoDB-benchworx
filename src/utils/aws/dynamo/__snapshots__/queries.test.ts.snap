// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Queries describeTable returns the information for the table 1`] = `
Object {
  "$metadata": Object {
    "attempts": 1,
    "cfId": undefined,
    "extendedRequestId": undefined,
    "httpStatusCode": 200,
    "requestId": undefined,
    "totalRetryDelay": 0,
  },
  "Table": Object {
    "ArchivalSummary": undefined,
    "AttributeDefinitions": Array [
      Object {
        "AttributeName": "Id",
        "AttributeType": "S",
      },
      Object {
        "AttributeName": "PostedBy",
        "AttributeType": "S",
      },
      Object {
        "AttributeName": "ReplyDateTime",
        "AttributeType": "S",
      },
    ],
    "BillingModeSummary": undefined,
    "CreationDateTime": Date { NaN },
    "GlobalSecondaryIndexes": undefined,
    "GlobalTableVersion": undefined,
    "ItemCount": 4,
    "KeySchema": Array [
      Object {
        "AttributeName": "Id",
        "KeyType": "HASH",
      },
      Object {
        "AttributeName": "ReplyDateTime",
        "KeyType": "RANGE",
      },
    ],
    "LatestStreamArn": undefined,
    "LatestStreamLabel": undefined,
    "LocalSecondaryIndexes": Array [
      Object {
        "IndexArn": "arn:aws:dynamodb:us-east-1:322567890963:table/Reply/index/PostedBy-index",
        "IndexName": "PostedBy-index",
        "IndexSizeBytes": 324,
        "ItemCount": 4,
        "KeySchema": Array [
          Object {
            "AttributeName": "Id",
            "KeyType": "HASH",
          },
          Object {
            "AttributeName": "PostedBy",
            "KeyType": "RANGE",
          },
        ],
        "Projection": Object {
          "NonKeyAttributes": undefined,
          "ProjectionType": "KEYS_ONLY",
        },
      },
    ],
    "ProvisionedThroughput": Object {
      "LastDecreaseDateTime": undefined,
      "LastIncreaseDateTime": undefined,
      "NumberOfDecreasesToday": 0,
      "ReadCapacityUnits": 6,
      "WriteCapacityUnits": 5,
    },
    "Replicas": undefined,
    "RestoreSummary": undefined,
    "SSEDescription": undefined,
    "StreamSpecification": undefined,
    "TableArn": "arn:aws:dynamodb:us-east-1:322567890963:table/Reply",
    "TableId": "bd8de77e-1cfc-4c24-9c35-fd1c1eb178f5",
    "TableName": "Reply",
    "TableSizeBytes": 472,
    "TableStatus": "ACTIVE",
  },
}
`;
